instrumentation/servlet/servlet-3.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/servlet/v3_0/snippet/SnippetInjectingResponseWrapper.java

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.instrumentation.servlet.v3_0.snippet;
+
+import static io.opentelemetry.javaagent.instrumentation.servlet.v3_0.snippet.Injection.initializeInjectionStateIfNeeded;
+import static java.util.logging.Level.FINE;
+
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+import javax.servlet.ServletOutputStream;
+import javax.servlet.http.HttpServletResponse;
+import javax.servlet.http.HttpServletResponseWrapper;
+
+/**
+ * Notes on Content-Length: the snippet length is only added to the content length when injection
+ * occurs and the content length was set previously.
+ *
+ * <p>If the Content-Length is set after snippet injection occurs (either for the first time or is
+ * set again for some reason),we intentionally do not add the snippet length, because the
+ * application server may be making that call at the end of a request when it sees the request has
+ * not been submitted, in which case it is likely using the real length of content that has been
+ * written, including the snippet length.
+ */
+public class SnippetInjectingResponseWrapper extends HttpServletResponseWrapper {
+
+  private static final Logger logger = Logger.getLogger(HttpServletResponseWrapper.class.getName());
+
+  public static final String FAKE_SNIPPET_HEADER = "FAKE_SNIPPET_HEADER";
+
+  private static final int UNSET = -1;
+
+  // this is for Servlet 3.1 support
+  @Nullable private static final MethodHandle setContentLengthLongHandler = getMethodHandle();
+
+  private final String snippet;
+  private final int snippetLength;
+
+  private long contentLength = UNSET;
+
+  private SnippetInjectingPrintWriter snippetInjectingPrintWriter = null;
+
+  public SnippetInjectingResponseWrapper(HttpServletResponse response, String snippet) {
+    super(response);
+    this.snippet = snippet;
+    snippetLength = snippet.length();
+  }
+
+  @Override
+  public boolean containsHeader(String name) {
+    // override this function in order to make sure the response is wrapped
+    // but not wrapped twice
+    // we didn't use the traditional method req.setAttribute
+    // because async would set the original request attribute but didn't pass down the wrapped
+    // response
+    // then the response would never be wrapped again
```

Comment:
```suggestion
    // because async requests pass down their attributes, but don't pass down our wrapped response
    // and so we would see the presence of the attribute and think the response was already wrapped
    // when it really is not
```